### ISSUE TITLE
`Observer`s are updated with validation data in quantization aware training

### ISSUE THREAD

**Issue by manipopopo** [PyTorchLightning/pytorch-lightning#8507]
## 🐛 Bug

`Observer`s are updated (trained) with validation data in quantization aware training. Should `pytorch_lightning.callbacks.QuantizationAwareTraining` apply `torch.quantization.disable_observer` on `pl_module` in `on_validation_start` or `on_validation_epoch_start`?

### To Reproduce

```python
import torch
from torch import nn
from torch import optim
from torch.utils import data
import pytorch_lightning as pl


class BoringModel(pl.LightningModule):

  def __init__(self):
    super().__init__()
    self.layer = nn.Conv2d(1, 1, (1, 1))

  def forward(self, x):
    return self.layer(x)

  def training_step(self, batch, batch_idx):
    raise NotImplementedError

  def validation_step(self, batch, batch_idx):
    self(batch)

    output_observer = self.layer.activation_post_process.activation_post_process
    # AssertionError.
    assert all(
        torch.isinf(val)
        for val in [output_observer.min_val, output_observer.max_val]
    )

  def configure_optimizers(self):
    return optim.SGD(self.parameters(), lr=0.1)


class RandomDataset(data.IterableDataset):

  def __iter__(self):
    yield from torch.rand(32, 1, 1, 1)

train_data_loader = data.DataLoader(RandomDataset(), batch_size=32)
val_data_loader = data.DataLoader(RandomDataset(), batch_size=32)

model = BoringModel()
trainer = pl.Trainer(
    num_sanity_val_steps=1,
    callbacks=[pl.callbacks.QuantizationAwareTraining()]
)
trainer.fit(
    model, train_dataloader=train_data_loader, val_dataloaders=val_data_loader
)
```

### Expected behavior

No `AssertionError` occurs. The observer should not be trained with validation data.

### Environment

 - PyTorch Lightning Version (e.g., 1.3.0): 1.3.8
 - PyTorch Version (e.g., 1.8): 1.9
 - Python version: 3.7.3
 - OS (e.g., Linux): Windows
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - How you installed PyTorch (`conda`, `pip`, source): `pip`
 - If compiling from source, the output of `torch.__config__.show()`:
 - Any other relevant information:

**Comment by Borda** [comment]
> Should `pytorch_lightning.callbacks.QuantizationAwareTraining` apply `torch.quantization.disable_observer` on `pl_module` in `on_validation_start` or `on_validation_epoch_start`?

that is a good question and I feel that users may want both... so how about disable the validation data usage by default but still add an argument for enabling it? so we would implement it in `on_validation_start`/`on_validation_end`
@PyTorchLightning/core-contributors @awaelchli other thoughts?
@manipopopo interested in sending such PR?

**Comment by manipopopo** [comment]
@Borda I'm glad to send a PR if the issue is confirmed.

**Comment by kandluis** [comment]
Yeah, this is a good issue. Most users would probably want to avoid calibrating with the validation set. My 2 cents here is that the model used for validation be as close as possible to the "final" model. From that POV, would it make sense to `convert` the model as well before the validation data is run? 

That makes the design more complex, but would also make the validation metrics more accurate.

**Comment by Borda** [comment]
> That makes the design more complex, but would also make the validation metrics more accurate.

so the conclusion is to add an argument that would disable observing for validation, right?

**Comment by manipopopo** [comment]
> would it make sense to convert the model as well before the validation data is run

It would be nice if we can performance validation with converted models. 

But I'm not sure how to properly handle the prepared model and the converted model in current `pytorch_lightning.callbacks.Callback`. And maybe the validation loop would be slow as we move the converted model to CPU to run the quantized operators.

Maybe we could only disable the observers for the moment?

```python

class QuantizationAwareTraining(Callback):

    def __init__(
        self,
        # ...
        val_disable_observers=True,
        test_disable_observers=True,
        predict_disable_observers=True,
    ):
        # ...
```

**Comment by Borda** [comment]
> Maybe we could just disable the observers for the moment?

@manipopopo lets split it, lest make one with the observer:
```py
enable_observers: Sequence[str] = ('train', 'valid')
```
so just list the phases which are used for collection stats, but we shall never use it for testing

Another PR would be adding an argument to run validation on the quantized model, but that shall be also on GPU if available

**Comment by kandluis** [comment]
> > That makes the design more complex, but would also make the validation metrics more accurate.
> 
> so the conclusion is to add an argument that would disable observing for validation, right?

Yes, I think this seems reasonable. I would say that by default we can disable calibration on the validation set as suggested above to keep this PR targetted.
